fontique/src/backend/coretext.rs

```diff
@@ -77,97 +71,39 @@ impl SystemFonts {
     pub(crate) fn fallback(&mut self, key: impl Into<FallbackKey>) -> Option<FamilyId> {
         let key = key.into();
         let sample = key.script().sample()?;
-        self.fallback_for_text(sample, key.locale(), false)
+        let font = self.fallback_font_for_text(sample, key.locale(), false)?;
+        let family_name = unsafe { CTFontCopyFamilyName(&font) };
+        self.name_map.get(&family_name.to_string()).map(|n| n.id())
     }
-}
 
-impl SystemFonts {
-    fn fallback_for_text(
+    fn fallback_font_for_text(
         &mut self,
         text: &str,
         locale: Option<&str>,
         prefer_ui: bool,
-    ) -> Option<FamilyId> {
-        let font = fallback_for_text(text, locale, prefer_ui)?;
-        self.name_map.get(&font.family_name()).map(|n| n.id())
-    }
-}
-
-fn fallback_for_text(text: &str, locale: Option<&str>, prefer_ui: bool) -> Option<CTFont> {
-    let cf_text = CFString::new(text);
-    let cf_text_range = CFRange::init(0, cf_text.char_len());
-    let cf_locale = locale.map(CFString::new);
-    let base_font = {
-        let desc_attrs = CFDictionary::from_CFType_pairs(&[]);
-        let mut desc = font_descriptor::new_from_attributes(&desc_attrs);
-        if prefer_ui {
-            if let Some(ui_desc) = ui_font_for_language(kCTFontSystemFontType, 0.0, None)
-                .map(|font| font.copy_descriptor())
-            {
-                desc = ui_desc;
-            }
-        }
-        font::new_from_descriptor(&desc, 0.0)
-    };
-    let font = unsafe {
-        CTFont::wrap_under_create_rule(if let Some(locale) = cf_locale {
-            CTFontCreateForStringWithLanguage(
-                base_font.as_concrete_TypeRef(),
-                cf_text.as_concrete_TypeRef(),
-                cf_text_range,
-                locale.as_concrete_TypeRef(),
-            )
-        } else {
-            CTFontCreateForString(
-                base_font.as_concrete_TypeRef(),
-                cf_text.as_concrete_TypeRef(),
-                cf_text_range,
-            )
-        })
-    };
-    Some(font)
-}
-
-fn ui_font_for_language(
-    ui_type: CTFontUIFontType,
-    size: f64,
-    language: Option<CFString>,
-) -> Option<CTFont> {
-    unsafe {
-        let font_ref = CTFontCreateUIFontForLanguage(
-            ui_type,
-            size,
-            language
-                .as_ref()
-                .map(|x| x.as_concrete_TypeRef())
-                .unwrap_or(std::ptr::null()),
-        );
-        if font_ref.is_null() {
-            None
+    ) -> Option<CFRetained<CTFont>> {
+        let text = CFString::from_str(text);
+        let text_range = CFRange {
+            location: 0,
+            length: unsafe { CFStringGetLength(&text) },
+        };
+        let locale = locale.map(CFString::from_str);
+        let base_font = if prefer_ui {
+            unsafe { CTFontCreateUIFontForLanguage(CTFontUIFontType::System, 0.0, None).unwrap() }
```

Comment:
I assume this can't fail. I could be wrong?

Reply:
I mean, since we are giving no `language` then presumably it is never going to fail in practice... though it is technically fallible.

Reply:
I believe you always get the last resort font when matching fails but I try to avoid unwraps in cases where I can’t be 100% certain. My general philosophy is that libraries should never panic, particularly in cases like this where the user has no control over the behavior.

Reply:
I will adjust this to remove the `unwrap`.